Content.Shared/Weapons/Hitscan/Systems/HitscanReflectSystem.cs

```diff
@@ -0,0 +1,48 @@
+using Content.Shared.Weapons.Hitscan.Components;
+using Content.Shared.Weapons.Hitscan.Events;
+using Content.Shared.Weapons.Ranged.Events;
+using Content.Shared.Weapons.Reflect;
+using Robust.Shared.Random;
+
+namespace Content.Shared.Weapons.Hitscan.Systems;
+
+public sealed class HitscanReflectSystem : EntitySystem
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<HitscanReflectComponent, HitscanRaycastFiredEvent>(OnHitscanHit);
+    }
+
+    private void OnHitscanHit(Entity<HitscanReflectComponent> hitscan, ref HitscanRaycastFiredEvent args)
+    {
+        if (hitscan.Comp.ReflectiveType == ReflectType.None || args.HitEntity == null)
+            return;
+
+        if (hitscan.Comp.CurrentReflections >= hitscan.Comp.MaxReflections)
+            return;
+
+        var ev = new HitScanReflectAttemptEvent(args.Shooter ?? args.Gun, args.Gun, hitscan.Comp.ReflectiveType, args.ShotDirection, false);
+        RaiseLocalEvent(args.HitEntity.Value, ref ev);
+
+        if (!ev.Reflected)
+            return;
+
+        hitscan.Comp.CurrentReflections++;
+
+        args.Canceled = true;
+
+        var fromEffect = Transform(args.HitEntity.Value).Coordinates;
+
+        var hitFiredEvent = new HitscanTraceEvent
```

Comment:
I don't think it's worth fixing (unless you want to) but I just realized that this resets the distance traveled on each reflect, meaning you can have the distance traveled by the raycast be quite a lot more than the intended max if you get a bunch of long distance reflects.